include/dnscpp/udps.h

```diff
@@ -0,0 +1,263 @@
+/**
+ *  Udp.h
+ *
+ *  Internal class that implements a UDP socket over which messages
+ *  can be sent to nameservers. You normally do not have to construct
+ *  this class in user space, it is used internally by the Context class.
+ *
+ *  @author Emiel Bruijntjes <emiel.bruijntjes@copernica.com>
+ *  @copyright 2020 - 2021 Copernica BV
+ */
+
+/**
+ *  Include guard
+ */
+#pragma once
+
+/**
+ *  Dependencies
+ */
+#include <vector>
+#include <algorithm>
+#include <stdlib.h>
+#include <sys/socket.h>
+#include "monitor.h"
+#include "watchable.h"
+#include "inbound.h"
+#include <list>
+#include <string>
+
+/**
+ *  Begin of namespace
+ */
+namespace DNS {
+
+/**
+ *  Forward declarations
+ */
+class Core;
+class Query;
+class Loop;
+class Ip;
+class Response;
+class Processor;
+
+/**
+ *  Class definition
+ */
+class Udps : private Watchable
+{
+public:
+    /**
+     *  Interface to be implemented by the parent
+     */
+    class Handler
+    {
+    public:
+        /**
+         *  Method that is called when the udp socket has a buffer available
+         *  @param  udp     the reporting socket
+         */
+        virtual void onBuffered(Udps *udp) = 0;
+    };
+
+private:
+    /**
+     *  event loop
+     *  @var Loop*
+     */
+    Loop *_loop;
+
+    /**
+     *  The object that is interested in handling responses
+     *  @var Handler*
+     */
+    Handler *_handler;
+
+    /**
+     *  Private helper struct to represent a socket
+     */
+    struct Socket final : private Monitor, private Inbound
+    {
+        /**
+         *  Pointer to a Udp object
+         *  @var Udp*
+         */
+        Udps *parent = nullptr;
```

Comment:
Define a private Handler interface that is to be implemented by the parent class instead of creating having reference to parent